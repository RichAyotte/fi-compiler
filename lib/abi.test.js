/**
 * @overview    ABI unit tests
 */

'use strict'

const fs = require('fs').promises

const compiler = require('../index')
const abi = require('./abi')

// Jest can't do better than this at the moment for async setup.
// Had to resort to https://github.com/facebook/jest/issues/1256#issuecomment-230996710
let compiledContract = null

beforeAll(async done => {
	const testContract = await fs.readFile('e2e/fi-samples/input_test.fi', {
		encoding: 'utf-8'
	})
	compiledContract = compiler.compile(testContract)
	done()
})

describe('ABI', () => {
	it(`loads valid ABI`, () => {
		expect(() => {
			abi.load(compiledContract.abi)
		}).not.toThrow()
		expect(compiledContract.abi).toMatchSnapshot()
	})

	it(`entry returns packed bytes`, () => {
		expect(abi.entry(`TestInputSimple`, {
			boolLabel: 'true'
			, natLabel: 1234
			, intLabel: -5678
			, stringLabel: `string`
			, mutezLabel: 1000
			, addressLabel: `address`
			, keyLabel: `key`
			, pkhLabel: `pkh`
			, signatureLabel: `signature`
			, bytesLabel: `bytes`
		})).toMatchSnapshot()

		// Complex input has not been implemented. Test should fail.
/*		expect(abi.entry(`TestInputComplex`, {
			mapLabel: {
				a: 1
				, b: 2
				, c: 3
			}
			, listLabel: [
				'string1'
				, 'string2'
				, 'string3'
			]
			, setLabel: [
				1
				, 2
				, 3
			]
		})).toMatchSnapshot()
*/	})

	it(`returns storage`, () => {
		const data = 'sample data'
		expect(abi.storage(data)).toBe(data)
	})
})
