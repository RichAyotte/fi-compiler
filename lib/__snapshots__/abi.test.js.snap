// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABI entry returns packed bytes 1`] = `"0x3715546e050707030a0707009213070700ee5807070100000006737472696e67070700a80f0707010000000761646472657373070701000000036b657907070100000003706b68070701000000097369676e61747572650369"`;

exports[`ABI loads valid ABI 1`] = `"{\\"storage\\":[{\\"name\\":\\"test\\",\\"type\\":[\\"nat\\"]}],\\"entry\\":[{\\"name\\":\\"TestInputSimple\\",\\"id\\":\\"0x3715546e\\",\\"input\\":[{\\"name\\":\\"boolLabel\\",\\"type\\":[\\"bool\\"]},{\\"name\\":\\"natLabel\\",\\"type\\":[\\"nat\\"]},{\\"name\\":\\"intLabel\\",\\"type\\":[\\"int\\"]},{\\"name\\":\\"stringLabel\\",\\"type\\":[\\"string\\"]},{\\"name\\":\\"mutezLabel\\",\\"type\\":[\\"mutez\\"]},{\\"name\\":\\"addressLabel\\",\\"type\\":[\\"address\\"]},{\\"name\\":\\"keyLabel\\",\\"type\\":[\\"key\\"]},{\\"name\\":\\"pkhLabel\\",\\"type\\":[\\"key_hash\\"]},{\\"name\\":\\"signatureLabel\\",\\"type\\":[\\"signature\\"]},{\\"name\\":\\"bytesLabel\\",\\"type\\":[\\"bytes\\"]}],\\"temp\\":[],\\"code\\":[[\\"set\\",\\"storage.test\\",[\\"input.natLabel\\"]]]}]}"`;
